ir/src/constraints/graph.rs

```diff
@@ -287,17 +287,44 @@ impl AlgebraicGraph {
         symbol_table: &SymbolTable,
         variable_roots: &mut VariableRoots,
         domain: ConstraintDomain,
+        scope: Scope,
         variable_value: VariableValue,
         variable_expr: &Expression,
     ) -> Result<ExprDetails, SemanticError> {
-        if let Some(expr) = variable_roots.get(&variable_value) {
-            // If the variable has already been inserted, return the existing expression details.
-            Ok(*expr)
-        } else {
-            // Otherwise, insert the variable expression and create a new variable root.
-            let expr = self.insert_expr(symbol_table, variable_expr, variable_roots, domain)?;
-            variable_roots.insert(variable_value, expr);
-            Ok(expr)
+        // The scope of the variable must be valid for the constraint domain.
+        match (scope, domain) {
+            (
+                Scope::BoundaryConstraints,
+                ConstraintDomain::FirstRow | ConstraintDomain::LastRow,
+            )
+            | (
+                Scope::IntegrityConstraints,
+                ConstraintDomain::EveryRow | ConstraintDomain::EveryFrame(_),
+            ) => {
+                let key = (scope, variable_value);
+                if let Some(expr) = variable_roots.get(&key) {
+                    match scope {
+                        Scope::BoundaryConstraints => {
+                            // TODO: deal with boundary conflict properly
+                            Ok(ExprDetails::new(
+                                expr.root_idx(),
+                                expr.trace_segment(),
+                                domain,
+                            ))
+                        }
+                        Scope::IntegrityConstraints => Ok(*expr),
+                    }
```

Comment:
This can be addressed during #150, but it actually requires significant refactoring to do it cleanly